### PR TITLE
Set up `style` mutation monitoring

### LINKED ISSUE #2008: Stop defensively copying style object in DOMComponent._updateDOMProperties
Right now, we copy it in one of the hottest paths of React to allow users to mutate the style object afterward. I've come to realize that this pattern doesn't come up _that_ often, since the style's often created inline inside `render` and not retained after that.

Additionally, `Object.assign` is coming soon, so people could always clone the style themselves if there's the need.

@jordwalke

### PR DESCRIPTION
**Pull request by chenglou**
Fixes #2008
This works by storing the style and its copy onto the component, and do the appropriate comparison during `_updateDOMProperties`.

### CODE REVIEW

**Comment by spicyj** [comment]
@chenglou lgtm – can you rebase, change console.warn to warning, and remove the monitorCodeUse?

**Comment by spicyj** [comment]
Also make the warning message use %s instead of manual concatenation – we're aiming to have all of the warning strings be static format strings.

**Comment by jsfb** [comment]
@chenglou ping.

**Comment by chenglou** [comment]
Sorry for the wait. Busy week.

File: src/browser/ui/ReactDOMComponent.js [hunk 1]
```diff
@@ -152,7 +196,14 @@ ReactDOMComponent.Mixin = {
       } else {
         if (propKey === STYLE) {
           if (propValue) {
+            if (__DEV__) {
+              // See `_updateDOMProperties`.
+              this._previousStyle = propValue;
+            }
             propValue = props.style = merge(props.style);
```
Comment (syranide) [inline src/browser/ui/ReactDOMComponent.js]: Should use `assign({}, props.style)` I think?

**Comment by chenglou** [comment]
I updated the PR but I think currently it's not possible for me to get the owner name for the warning (`_currentElement._owner` is always null for DOMElement). Otherwise it's fine (warning message not as useful though).

@zpao

**Comment by spicyj** [comment]
@chenglou if you rebase the owner thing should work.

**Comment by chenglou** [comment]
Done!

**Comment by chenglou** [comment]
Merging then.

**Pull request merged**